cpp/src/request_am.cpp

```diff
@@ -230,59 +233,95 @@ ucs_status_t RequestAm::recvCallback(void* arg,
   }
 }
 
-std::shared_ptr<Buffer> RequestAm::getRecvBuffer() { return _buffer; }
+std::shared_ptr<Buffer> RequestAm::getRecvBuffer()
+{
+  return std::visit(data::dispatch{
+                      [](data::AmReceive amReceive) { return amReceive._buffer; },
+                      [](auto arg) {
+                        throw std::runtime_error("Unreachable");
+                        return decltype(data::AmReceive::_buffer){nullptr};
+                      },
```

Comment:
Done in ffc1c10d, but specifying the return type directly. Attempting that we get errors such as below because this is not a static member:

```
/repo/cpp/src/request_am.cpp: In member function 'virtual std::shared_ptr<ucxx::Buffer> ucxx::RequestAm::getRecvBuffer()':
/repo/cpp/src/request_am.cpp:237:73: error: invalid use of non-static member function 'virtual std::shared_ptr<ucxx::Buffer> ucxx::RequestAm::getRecvBuffer()'
  237 |       [](auto) -> std::invoke_result_t<decltype(RequestAm::getRecvBuffer)> { throw std::runtime_error("Unreachable"); },
      |                                                                         ^
/repo/cpp/src/request_am.cpp:232:25: note: declared here
  232 | std::shared_ptr<Buffer> RequestAm::getRecvBuffer()
      |                         ^~~~~~~~~
/repo/cpp/src/request_am.cpp:237:74: error: template argument 1 is invalid
  237 |       [](auto) -> std::invoke_result_t<decltype(RequestAm::getRecvBuffer)> { throw std::runtime_error("Unreachable"); },
      |                                  
```

I know how to fix it with `std::declval` but I'm not sure if using it is a great solution. This is such a simple case that I don't think it's worth the time to find a different (and likely more verbose and confusing) approach anyway.